lib/dartdoc.dart

```diff
@@ -429,6 +451,34 @@ class Dartdoc extends PackageBuilder {
     _doOrphanCheck(packageGraph, origin, visited);
     _doSearchIndexCheck(packageGraph, origin, visited);
   }
+
+  /// Runs [generateDocs] function and properly handles the errors.
+  Future<int> execute() async {
+    onCheckProgress.listen(logProgress);
+    try {
+      await runZoned(generateDocs,
```

Comment:
Every time I've touched this area of code it seems I trade off one set of hard to understand behavior for another... I'd really like to believe you've nailed it this time.  I'm OK with this as long as you've verified that:

*  when exceptions are thrown in generateDocs when run in a debugger, the debugger stops correctly.  The reason why we didn't use unconditional catch was related to this IIRC and I found it too frustrating.
*  exception output is reasonable when we throw an exception in an async function.

Reply:
Can you let me know how to recreate those scenarios? I don't have these changes in my g3 client and I would like to be able to test it like you would have.

- I somehow need to compile and run dartdoc on a separate package I assume. How do I do that? I tried to inspect a few tests but I think you telling me is faster :-)
- How do I wire up the debugger? I am using VSCode. I can also switch to IntelliJ. 

Sorry, google3 engineers are total noobs when it comes to upstream development :) 

Or if it is faster, perhaps you can patch and try it out?

Reply:
Trust me I know the feeling, wound up here from g3 myself :-)

Running debug in VSCode should be the same as IntelliJ as far as I know -- essentially, define a simple execution of dartdoc in whatever VSCode's equivalent of the little pull down list of command lines to run IntelliJ has, and click "debug" for it.  This seems relevant:  https://code.visualstudio.com/docs/editor/debugging.  Having dartdoc document itself or the test package at `testing/test_package` is what I do for most simple cases.

If you have the right Dart plugin it should work out of the box, though I'm no expert.

I'll give this a shot myself later today if you don't get to it, should be pretty quick to test.

Reply:
Oh, right, how to document something!  Running the dartdoc binary with no parameters and the CWD in the root of the package you are documenting is the most basic way to run it.

Reply:
I tested this. In normal code (such as some random point in `Class`). You can see async gaps in both cases:

- Debugging works fine. Stack looks OK.
![debug](https://user-images.githubusercontent.com/6687929/81995676-cd319800-95ff-11ea-9e50-0773ec1f7941.png)

- Throwing an exception is fine:

```
dartdoc failed: NoSuchMethodError: The getter 'isEmpty' was called on null.
Receiver: null
Tried calling: isEmpty
#0      Object.noSuchMethod  (dart:core-patch/object_patch.dart:53:5)
#1      Class.inheritedOperators package:dartdoc/…/model/class.dart:196
#2      Class.instanceOperators package:dartdoc/…/model/class.dart:68
#3      Container.allModelElements package:dartdoc/…/model/container.dart:41
#4      Class.allModelElements package:dartdoc/…/model/class.dart:75
#5      Library.modelElementsMap.<anonymous closure> package:dartdoc/…/model/library.dart:582
#6      ExpandIterator.moveNext  (dart:_internal/iterable.dart:484:31)
#7      ExpandIterator.moveNext  (dart:_internal/iterable.dart:478:31)
#8      Iterable.forEach  (dart:core/iterable.dart:279:23)
#9      Library.modelElementsMap package:dartdoc/…/model/library.dart:599
#10     Library.allModelElements package:dartdoc/…/model/library.dart:614
#11     PackageGraph.allModelElements.<anonymous closure>.<anonymous closure>  package:dartdoc/…/model/package_graph.dart:874
#12     SetMixin.forEach  (dart:collection/set.dart:148:30)
#13     PackageGraph.allModelElements.<anonymous closure> package:dartdoc/…/model/package_graph.dart:873
#14     SetMixin.forEach  (dart:collection/set.dart:148:30)
#15     PackageGraph.allModelElements package:dartdoc/…/model/package_graph.dart:867
#16     PackageGraph.precacheLocalDocs.<anonymous closure>  package:dartdoc/…/model/package_graph.dart:115
#17     _SyncIterator.moveNext  (dart:core-patch/core_patch.dart:146:23)
#18     new List.from  (dart:core-patch/array_patch.dart:47:19)
#19     Iterable.toList  (dart:core/iterable.dart:400:12)
#20     PackageGraph.initializePackageGraph package:dartdoc/…/model/package_graph.dart:66
#21     PubPackageBuilder.buildPackageGraph package:dartdoc/…/model/package_builder.dart:67
<asynchronous suspension>
#22     Dartdoc.generateDocsBase package:dartdoc/dartdoc.dart:175
#23     Dartdoc.generateDocs package:dartdoc/dartdoc.dart:215
#24     _rootRun  (dart:async/zone.dart:1184:13)
#25     _CustomZone.run  (dart:async/zone.dart:1077:19)
#26     _runZoned  (dart:async/zone.dart:1619:10)
#27     runZoned  (dart:async/zone.dart:1539:10)
#28     Dartdoc.execute  package:dartdoc/dartdoc.dart:459
#29     main 
```

Reply:
I also tested it directly in an async function and it also works fine.

I added a comment to that code block so folks don't change it back blindly. I would love to get a concrete case where debugging is problematic.

I tracked the original change to `Chain` to a PR submitted 5 years ago: https://github.com/dart-lang/dartdoc/commit/e01acac8a5c722956079bb101b66eaf06c512a79. The code looks very different now and we are using Zones. I think my change is a step in the right direction.